File: 20-valid-parentheses/solution.cpp
#include <string>
#include <deque>
#include <unordered_set>

class Solution
{
private:
  static bool paren_match(const char starter, const char closer)
  {
    return (starter == '(' && closer == ')') || (starter == '[' && closer == ']') || (starter == '{' && closer == '}');
  }

public:
  bool isValid(std::string s)
  {
    if (s.size() % 2 != 0)
    {
      return false;
    }

    std::unordered_set<char> starters{'(', '[', '{'};
    std::deque<char> stack;
    for (const char curr : s)
    {
      if (starters.contains(curr))
      {
        stack.emplace_back(curr);
        continue;
      }

      if (stack.empty())
      {
        return false;
      }

      const char prev = stack.back();
      stack.pop_back();
      if (!Solution::paren_match(prev, curr))
      {
        return false;
      }
    }

    return stack.empty();
  }
};

File: 20-valid-parentheses/solution.py
from collections import deque


class Solution:
  @staticmethod
  def paren_match(starter: str, closer: str) -> bool:
    return (starter == '(' and closer == ')') or (starter == '[' and closer == ']') or (starter == '{' and closer == '}')

  def isValid(self, s: str) -> bool:
    if len(s) % 2 != 0:
      return False

    starters = {'(', '[', '{'}
    stack: deque[str] = deque()
    for curr in s:
      if curr in starters:
        stack.append(curr)
        continue

      if not stack:
        return False

      prev = stack.pop()
      if not Solution.paren_match(prev, curr):
        return False

    return not stack
